src/WebJobs.Script.WebHost/AssemblyAnalysis/AssemblyAnalysisService.cs

```diff
@@ -0,0 +1,158 @@
+            _scriptHost = scriptHost;
+            _loggerFactory = loggerFactory;
+            _standbyOptionsMonitor = standbyOptionsMonitor;
+        }
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (!_environment.IsCoreTools())
+            {
+                if (_standbyOptionsMonitor.CurrentValue.InStandbyMode)
+                {
+                    _standbyOptionsMonitor.OnChange(standbyOptions =>
+                    {
+                        if (!standbyOptions.InStandbyMode && !_analysisScheduled)
+                        {
+                            ScheduleAssemblyAnalysis();
+                        }
+                    });
+                }
+                else
+                {
+                    ScheduleAssemblyAnalysis();
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            _cancellationTokenSource?.Cancel();
+
+            if (_analysisTask != null && !_analysisTask.IsCompleted)
+            {
+                var logger = _loggerFactory.CreateLogger<AssemblyAnalysisService>();
+                logger.LogDebug("Assembly analysis service stopped before analysis completion. Waiting for cancellation.");
+
+                return _analysisTask;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void ScheduleAssemblyAnalysis()
+        {
+            var jobHost = _scriptHost.GetService<IScriptJobHost>();
+            if (jobHost == null
+                || !jobHost.Functions.Any(f => f.Metadata.IsDirect()))
+            {
+                return;
+            }
+
+            _analysisScheduled = true;
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            _analysisTask = Task.Delay(TimeSpan.FromMinutes(1), _cancellationTokenSource.Token)
+               .ContinueWith(t => AnalyzeFunctionAssemblies());
+        }
+
+        private void AnalyzeFunctionAssemblies()
+        {
+            var jobHost = _scriptHost.GetService<IScriptJobHost>();
+
+            if (jobHost == null
+                || _cancellationTokenSource.IsCancellationRequested)
+            {
+                return;
+            }
+
+            var logger = _loggerFactory.CreateLogger<AssemblyAnalysisService>();
+            var assemblies = new HashSet<Assembly>();
+            var hasUnoptimizedAssemblies = false;
+
+            foreach (var item in jobHost.Functions)
+            {
+                if (_cancellationTokenSource.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                if (item.Metadata.Properties.TryGetValue(ScriptConstants.FunctionMetadataDirectTypeKey, out Type type)
+                    && !assemblies.Contains(type.Assembly))
+                {
+                    if (!IsReadyToRun(type.Assembly))
```

Comment:
There is no need to loop thru all Functions, if one is not optimized the rest will be not optimized as well.

I suggest we exit the loop if/when we detected one and logged it. that's enough for Risk Alerts.

Reply:
For most applications, that will be the case, but not all functions have to come from the same assembly, and although it’s probably safe to assume that in the normal publishing flow, either all assemblies defining functions would be either be optimized or all of them wouldn’t, there are scenarios where that could deviate from that, though, but pretty unlikely ones.

I agree with you that for the risk alerts, simply flagging that we have unoptimized assemblies and exiting would be sufficient to meet our requirements, and debated doing that, but ended up doing this as I thought it would be helpful to also identify all of the assemblies and provide that information. In the vast majority of cases, we’ll be dealing with a single assembly for all functions, so the actual analysis would only be performed once anyway, and given the average number of functions per app and how this runs, I don’t believe we would see significant impact if we exit early, but either way is really fine for our purposes. 

I don’t think the current approach is a problem, and it might be good to list all assemblies, but I’m happy to change to simply flag the app, as I agree this is not super important. Thoughts?

Reply:
ok sounds good.